### PR TITLE
feat(check): verify the metadata-schema and account-type projections

### PR DESCRIPTION
**Pull request by Azorlogh**
> **Stacked on #1554** (`fix/rebuilddelta-schema-2`) — review/merge that first.

## Problem

#1554 makes `RebuildDelta` reconstruct the metadata schema on restore, but the
checker never *verified* the `LedgerInfo` projections (`MetadataSchema`,
`AccountTypes`) — invariant 8 requires a compare pass for each. Closing that gap
surfaced two deeper issues: the checker seeded some *expected* state from the
**live store** (circular — the store checked against itself), and it never
verified the **ledger set** itself.

## Changes

- **Verify the projections.** The baseline snapshot now carries `LedgerInfo`, an
  independent boundary-time source. `compareSchema` re-derives the schema
  (`CreateLedger.initial_schema` + `SetMetadataFieldType`/`RemovedMetadataFieldType`)
  → `SCHEMA_MISMATCH`; `compareAccountTypes` re-derives account types
  (`CreateLedger.account_types` + `Add`/`RemoveAccountType`) → `ACCOUNT_TYPE_MISMATCH`.

- **No checker state is seeded from the live store.** The baseline is opened once,
  up front, and is the sole source of all pre-archive state — the live ledger set,
  schema, account types, and reversion tracking. (Also fixes a purge-sim timing bug
  where account types were seeded from their own end state.) Archived-but-no-baseline
  (e.g. checkpoint-only restore) skips entry-by-entry verification, logged at **error**
  level; the audit hash chain is still verified.

- **Verify the ledger set both ways.** `compareLedgerPresence`: an audit-live ledger
  with no live stored `LedgerInfo` (missing, or tampered to a tombstone) →
  `MISSING_LEDGER`; a stored `LedgerInfo` with no `CreateLedger` in the audit →
  `UNAUDITED_LEDGER`.

- **e2e.** `CheckStore` over schema + account types: normal store, account types at
  creation, and a chapter-archive boundary (pre-archive declarations live only in the
  baseline).

New error types: `SCHEMA_MISMATCH` (11), `ACCOUNT_TYPE_MISMATCH` (12),
`MISSING_LEDGER` (13), `UNAUDITED_LEDGER` (14).

## Verification

Unit tests cover each compare pass (match / missing / tampered / unaudited /
pending-cleanup) and the baseline round-trip; the focused `CheckStore` e2e incl. the
archive-boundary case passes. `go build ./...`, `go vet`, `golangci-lint` clean on
touched packages. Detection is unit-tested — provoking it end-to-end needs corrupting
the stored `LedgerInfo`, unreachable via the API.

### CODE REVIEW

File: internal/application/check/checker.go [hunk 1]
```diff
@@ -345,12 +364,16 @@ func (c *Checker) Check(ctx context.Context, callback func(*servicepb.CheckStore
 			switch payload := log.GetPayload().GetType().(type) {
 			case *commonpb.LogPayload_CreateLedger:
 				if payload.CreateLedger != nil {
-					knownLedgers[payload.CreateLedger.GetName()] = struct{}{}
+					name := payload.CreateLedger.GetName()
```
Comment (NumaryBot) [inline internal/application/check/checker.go:352]: 🟠 **[major] Seed initial account types from CreateLedger**

When a ledger is created with `CreateLedgerRequest.account_types`, `processCreateLedger` stores and logs those initial types, but this replay branch only seeds `expectedSchemas`. A valid store for such a ledger will later compare `expectedAccountTypes[name] == nil` against `LedgerInfo.AccountTypes` and emit `ACCOUNT_TYPE_MISMATCH` even though nothing is corrupt. Please initialize the expected account-type map from `payload.CreateLedger.GetAccountTypes()` here.

File: internal/application/check/checker.go [hunk 2]
```diff
@@ -1935,6 +1982,242 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	})
+	if err != nil {
+		c.logger.Infof("failed to open baseline checkpoint for seeding: %v", err)
+
+		return
+	}
+
+	defer func() { _ = db.Close() }()
+
+	ledgerCursor, err := query.ReadLedgers(ctx, db)
```
Comment (NumaryBot) [inline internal/application/check/checker.go]: 🟠 **[major] Handle legacy baselines without LedgerInfo**

On nodes that already have a baseline checkpoint created by the previous code, this read succeeds but returns no ledgers because old baselines only contain attribute zones. Since `baselineDB != nil` later allows `compareSchema`/`compareAccountTypes` to run, any schema or account types declared before the archive are missing from `expected` and valid stores will be reported as mismatches after upgrade. Please detect an unseeded/legacy baseline and skip or otherwise handle these projection checks for archived chapters.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 2 new inline findings.

Summary: https://github.com/formancehq/ledger/pull/1555#issuecomment-4928117739

File: internal/application/check/checker.go [hunk 3]
```diff
@@ -345,12 +358,17 @@ func (c *Checker) Check(ctx context.Context, callback func(*servicepb.CheckStore
 			switch payload := log.GetPayload().GetType().(type) {
 			case *commonpb.LogPayload_CreateLedger:
 				if payload.CreateLedger != nil {
-					knownLedgers[payload.CreateLedger.GetName()] = struct{}{}
+					name := payload.CreateLedger.GetName()
```
Comment (NumaryBot) [inline internal/application/check/checker.go:352]: 🟠 **[major] Initial account types from CreateLedger not seeded into expected map**

When a ledger is created with account types via `CreateLedgerRequest.account_types`, the replay branch initialises `expectedSchemas` but does not populate `expectedAccountTypes` from `payload.CreateLedger.GetAccountTypes()`. Later comparisons will therefore see `expectedAccountTypes[name] == nil` for every pre-existing account type and emit false-positive `ACCOUNT_TYPE_MISMATCH` errors even though the store is not corrupt.

**Suggestion:** In the `processCreateLedger` replay branch (around line 361), iterate over `payload.CreateLedger.GetAccountTypes()` and seed `expectedAccountTypes` with those values, mirroring how `expectedSchemas` is already seeded.

File: internal/application/check/checker.go [hunk 4]
```diff
@@ -1935,6 +1969,223 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	})
+	if err != nil {
+		c.logger.Infof("failed to open baseline checkpoint for seeding: %v", err)
+
+		return
+	}
+
+	defer func() { _ = db.Close() }()
+
+	ledgerCursor, err := query.ReadLedgers(ctx, db)
```
Comment (NumaryBot) [inline internal/application/check/checker.go]: 🟠 **[major] Legacy baselines without LedgerInfo cause false-positive mismatches on archived stores**
<sub>reported by NumaryBot, codex</sub>

Existing baseline checkpoints created before this PR only contain attribute zones and no `LedgerInfo`. When such a baseline is read, `baselineDB != nil` is still true, so `compareSchema` and `compareAccountTypes` run. Because account types and schemas declared before the archive point are absent from the `expected` maps, every such entry appears as a mismatch, producing false positives for valid stores after an upgrade.

**Suggestion:** After reading the baseline, check whether it contains a `LedgerInfo` section (or equivalent sentinel). If the section is absent, treat the baseline as a legacy checkpoint and skip the schema/account-type projection checks for archived chapters, or re-seed the expected maps from the baseline data before running comparisons.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 2 new inline findings.

Summary: https://github.com/formancehq/ledger/pull/1555#issuecomment-4928117739

File: internal/application/check/checker.go (hunk 3 above)
Comment (Azorlogh) [reply internal/application/check/checker.go:352]: https://github.com/formancehq/ledger/pull/1555/commits/1336f41b170f6c4196d9b524b9fe59eb7abbd34c

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 3 new inline findings.

Summary: https://github.com/formancehq/ledger/pull/1555#issuecomment-4928117739

File: internal/application/check/checker.go [hunk 5]
```diff
@@ -345,12 +358,18 @@ func (c *Checker) Check(ctx context.Context, callback func(*servicepb.CheckStore
 			switch payload := log.GetPayload().GetType().(type) {
 			case *commonpb.LogPayload_CreateLedger:
 				if payload.CreateLedger != nil {
-					knownLedgers[payload.CreateLedger.GetName()] = struct{}{}
+					name := payload.CreateLedger.GetName()
```
Comment (NumaryBot) [inline internal/application/check/checker.go:352]: 🟠 **[major] Initial account types from CreateLedger not seeded into expectedAccountTypes**
<sub>reported by NumaryBot, codex</sub>

When a ledger is created with account types via CreateLedgerRequest.account_types, the replay branch initialises expectedSchemas but does not populate expectedAccountTypes from payload.CreateLedger.GetAccountTypes(). Later comparisons therefore see expectedAccountTypes[name] == nil for every pre-existing account type and emit false-positive ACCOUNT_TYPE_MISMATCH errors even though the store is not corrupt.

**Suggestion:** In the processCreateLedger replay branch (around line 361), iterate over payload.CreateLedger.GetAccountTypes() and seed expectedAccountTypes with those values, mirroring how expectedSchemas is already seeded.

File: internal/application/check/checker.go [hunk 6]
```diff
@@ -1935,6 +1970,234 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	})
+	if err != nil {
+		c.logger.Infof("failed to open baseline checkpoint for seeding: %v", err)
+
+		return
+	}
+
+	defer func() { _ = db.Close() }()
+
+	ledgerCursor, err := query.ReadLedgers(ctx, db)
```
Comment (NumaryBot) [inline internal/application/check/checker.go]: 🟠 **[major] Legacy baselines without LedgerInfo cause false-positive mismatches on archived stores**
<sub>reported by NumaryBot, codex</sub>

Existing baseline checkpoints created before this PR only contain attribute zones and no LedgerInfo section. When such a baseline is read, baselineDB != nil is still true, so compareSchema and compareAccountTypes run. Because account types and schemas declared before the archive point are absent from the expected maps, every such entry appears as a mismatch, producing false positives for valid stores after an upgrade.

**Suggestion:** After reading the baseline, check whether it contains a LedgerInfo section (or equivalent sentinel). If the section is absent, treat the baseline as a legacy checkpoint and skip the schema/account-type projection checks for archived chapters, or re-seed the expected maps from the baseline data before running comparisons.

File: internal/application/check/checker.go [hunk 7]
```diff
@@ -1935,6 +1970,234 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	if err != nil {
+		return fmt.Errorf("reading ledgers for schema verification: %w", err)
+	}
+
+	ledgers, err := cursor.Collect(ledgerCursor)
+	if err != nil {
+		return fmt.Errorf("collecting ledgers for schema verification: %w", err)
+	}
+
+	for _, info := range ledgers {
```
Comment (NumaryBot) [inline internal/application/check/checker.go:1956]: 🟠 **[major] Missing LedgerInfo projections in stored ledger list not detected**

If a live ledger's LedgerInfo entry is deleted from Pebble while the audit log still contains its CreateLedger event (e.g. a ledger with only schema/account-type declarations and no volumes), the expected map contains that ledger but query.ReadLedgers will not return it. The current loop over stored ledgers therefore emits no schema or account-type mismatch, creating a blind spot in the integrity verification the PR is meant to provide.

**Suggestion:** After iterating over stored ledgers, iterate over the expected maps and report any ledger present in expected but absent from the stored ledger list as a missing LedgerInfo projection mismatch.

File: internal/application/check/checker.go [hunk 8]
```diff
@@ -345,12 +361,18 @@ func (c *Checker) Check(ctx context.Context, callback func(*servicepb.CheckStore
 			switch payload := log.GetPayload().GetType().(type) {
 			case *commonpb.LogPayload_CreateLedger:
 				if payload.CreateLedger != nil {
-					knownLedgers[payload.CreateLedger.GetName()] = struct{}{}
+					name := payload.CreateLedger.GetName()
```
Comment (NumaryBot) [inline internal/application/check/checker.go:352]: 🟠 **[major] Initial account types from CreateLedger not seeded into expectedAccountTypes**

When a ledger is created with account types via CreateLedgerRequest.account_types, the replay branch initialises expectedSchemas but does not populate expectedAccountTypes from payload.CreateLedger.GetAccountTypes(). Later comparisons therefore see expectedAccountTypes[name] == nil for every pre-existing account type and emit false-positive ACCOUNT_TYPE_MISMATCH errors even though the store is not corrupt.

**Suggestion:** In the processCreateLedger replay branch (around line 364), iterate over payload.CreateLedger.GetAccountTypes() and seed expectedAccountTypes with those values, mirroring how expectedSchemas is already seeded.

File: internal/application/check/checker.go [hunk 9]
```diff
@@ -1935,6 +1977,287 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	})
+	if err != nil {
+		c.logger.Infof("failed to open baseline checkpoint for seeding: %v", err)
+
+		return
+	}
+
+	defer func() { _ = db.Close() }()
+
+	ledgerCursor, err := query.ReadLedgers(ctx, db)
```
Comment (NumaryBot) [inline internal/application/check/checker.go]: 🟠 **[major] Legacy baselines without LedgerInfo cause false-positive mismatches on archived stores**

Existing baseline checkpoints created before this PR only contain attribute zones and no LedgerInfo section. When such a baseline is read, baselineDB != nil is still true, so compareSchema and compareAccountTypes run. Because account types and schemas declared before the archive point are absent from the expected maps, every such entry appears as a mismatch, producing false positives for valid stores after an upgrade.

**Suggestion:** After reading the baseline, check whether it contains a LedgerInfo section (or equivalent sentinel). If the section is absent, treat the baseline as a legacy checkpoint and skip the schema/account-type projection checks for archived chapters, or re-seed the expected maps from the baseline data before running comparisons.

File: internal/application/check/checker.go [hunk 10]
```diff
@@ -1935,6 +1977,287 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	if err != nil {
+		return fmt.Errorf("reading ledgers for schema verification: %w", err)
+	}
+
+	ledgers, err := cursor.Collect(ledgerCursor)
+	if err != nil {
+		return fmt.Errorf("collecting ledgers for schema verification: %w", err)
+	}
+
+	for _, info := range ledgers {
```
Comment (NumaryBot) [inline internal/application/check/checker.go:1956]: 🟠 **[major] Missing LedgerInfo projections in stored ledger list not detected**

If a live ledger's LedgerInfo entry is deleted from Pebble while the audit log still contains its CreateLedger event (e.g. a ledger with only schema/account-type declarations and no volumes), the expected map contains that ledger but query.ReadLedgers will not return it. The current loop over stored ledgers therefore emits no schema or account-type mismatch, creating a blind spot in the integrity verification the PR is meant to provide.

**Suggestion:** After iterating over stored ledgers, iterate over the expected maps and report any ledger present in expected but absent from the stored ledger list as a missing LedgerInfo projection mismatch.

File: internal/application/check/checker.go [hunk 11]
```diff
@@ -1935,6 +1977,287 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	if err != nil {
+		return fmt.Errorf("reading ledgers for presence verification: %w", err)
+	}
+
+	ledgers, err := cursor.Collect(ledgerCursor)
+	if err != nil {
+		return fmt.Errorf("collecting ledgers for presence verification: %w", err)
+	}
+
+	stored := make(map[string]struct{}, len(ledgers))
```
Comment (NumaryBot) [inline internal/application/check/checker.go]: 🟠 **[major] Soft-deleted LedgerInfo entries not treated as missing during projection check**

If a live ledger's LedgerInfo projection is corrupted such that DeletedAt is set, compareSchema and compareAccountTypes both skip it as a tombstone, but the presence check still records the name as stored. An audit-live ledger with a soft-deleted LedgerInfo therefore produces no MISSING_LEDGER, schema, or account-type mismatch, creating a blind spot for this specific tampering shape.

**Suggestion:** Only count a LedgerInfo entry as satisfying the knownLedgers check when DeletedAt is not set; treat soft-deleted entries as absent so they trigger the same mismatch path as a fully missing entry.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 4 new inline findings.

Summary: https://github.com/formancehq/ledger/pull/1555#issuecomment-4928117739

File: internal/application/check/checker.go [hunk 12]
```diff
@@ -1935,6 +1977,296 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	})
+	if err != nil {
+		c.logger.Infof("failed to open baseline checkpoint for seeding: %v", err)
+
+		return
+	}
+
+	defer func() { _ = db.Close() }()
+
+	ledgerCursor, err := query.ReadLedgers(ctx, db)
```
Comment (NumaryBot) [inline internal/application/check/checker.go]: 🟠 **[major] Legacy baselines without LedgerInfo cause false-positive mismatches on archived stores**
<sub>reported by NumaryBot, codex</sub>

Existing baseline checkpoints created before this PR contain no LedgerInfo section. When such a baseline is read, `baselineDB != nil` is still true, so `compareSchema` and `compareAccountTypes` run. Because account types and schemas declared before the archive point are absent from the expected maps, every such entry appears as a mismatch, producing false positives for valid stores after an upgrade.

**Suggestion:** After reading the baseline, check whether it contains a LedgerInfo section (or equivalent sentinel). If absent, treat the baseline as a legacy checkpoint and skip the schema/account-type projection checks for archived chapters, or re-seed the expected maps from the baseline data before running comparisons.

File: internal/application/check/checker.go [hunk 13]
```diff
@@ -1935,6 +1977,296 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	if err != nil {
+		return fmt.Errorf("reading ledgers for schema verification: %w", err)
+	}
+
+	ledgers, err := cursor.Collect(ledgerCursor)
+	if err != nil {
+		return fmt.Errorf("collecting ledgers for schema verification: %w", err)
+	}
+
+	for _, info := range ledgers {
```
Comment (NumaryBot) [inline internal/application/check/checker.go:1956]: 🟠 **[major] Missing LedgerInfo projections in stored ledger list not detected**

If a live ledger's LedgerInfo entry is absent from Pebble while its CreateLedger event remains in the audit log, the loop over stored ledgers emits no schema or account-type mismatch, creating a blind spot in the integrity verification the PR is meant to provide.

**Suggestion:** After iterating over stored ledgers, iterate over the expected maps and report any ledger present in expected but absent from the stored ledger list as a missing LedgerInfo projection mismatch.

File: internal/application/check/checker.go [hunk 14]
```diff
@@ -1935,6 +1977,296 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	ledgerCursor, err := query.ReadLedgers(ctx, reader)
+	if err != nil {
+		return fmt.Errorf("reading ledgers for presence verification: %w", err)
+	}
+
+	ledgers, err := cursor.Collect(ledgerCursor)
+	if err != nil {
+		return fmt.Errorf("collecting ledgers for presence verification: %w", err)
+	}
+
```
Comment (NumaryBot) [inline internal/application/check/checker.go:2120]: 🟠 **[major] Soft-deleted LedgerInfo entries not treated as missing during projection check**

If a live ledger's LedgerInfo projection is corrupted such that DeletedAt is set, `compareSchema` and `compareAccountTypes` both skip it as a tombstone, but the presence check still records the name as stored. An audit-live ledger with a soft-deleted LedgerInfo therefore produces no MISSING_LEDGER, schema, or account-type mismatch, creating a blind spot for this specific tampering shape.

**Suggestion:** Only count a LedgerInfo entry as satisfying the knownLedgers check when DeletedAt is not set; treat soft-deleted entries as absent so they trigger the same mismatch path as a fully missing entry.

File: internal/application/check/checker.go [hunk 15]
```diff
@@ -1935,6 +1977,296 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	live := make(map[string]struct{}, len(ledgers))
+	for _, info := range ledgers {
+		if info.GetDeletedAt() != nil {
+			continue
+		}
+
+		live[info.GetName()] = struct{}{}
+	}
+
+	for name := range knownLedgers {
```
Comment (NumaryBot) [inline internal/application/check/checker.go:2139]: 🟠 **[major] Unaudited live LedgerInfo rows not detected (audit ↔ store asymmetry)**

If Pebble contains a live LedgerInfo for a ledger that never appears in the audit log (e.g. a tampered row with no schema or account types), `compareSchema` and `compareAccountTypes` treat empty stored projections as equal to empty expected values. The presence check only looks for audit-known ledgers missing from the live set, so the inverse — live rows absent from the audit — passes undetected. `CheckStore` can therefore pass while `ListLedgers` exposes an unaudited ledger.

**Suggestion:** After the audit-vs-store loop, also iterate over stored live ledger rows and report any that are absent from the audit-derived expected set as an integrity mismatch, taking care not to seed that set from the live store on archived paths.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 4 new inline findings.

Summary: https://github.com/formancehq/ledger/pull/1555#issuecomment-4928117739

File: internal/application/check/checker.go [hunk 16]
```diff
@@ -230,58 +242,50 @@ func (c *Checker) Check(ctx context.Context, callback func(*servicepb.CheckStore
...
+			return nil
 		}
 
-		if err := txIter.Close(); err != nil {
-			return fmt.Errorf("closing tx streaming iter: %w", err)
-		}
+		// The pre-archive live ledger set, schema, and account types were declared
+		// in now-purged logs; seed them from the boundary-time baseline so the
+		// post-archive replay applies its delta on top.
+		c.seedExpectedFromBaseline(ctx, baselineDB, knownLedgers, expectedSchemas, rawLedgerTypes, ledgerAccountTypes)
```
Comment (NumaryBot) [inline internal/application/check/checker.go:283]: 🔴 **[blocker] Legacy baselines without LedgerInfo cause false-positive mismatches on archived stores**
<sub>reported by NumaryBot, codex</sub>

Baseline checkpoints created before this PR contain only attribute zones and no LedgerInfo section. When such a baseline is opened, `baselineDB != nil` is still true, so `compareSchema` and `compareAccountTypes` run. Because schemas and account types declared before the archive point are absent from the expected maps (since ReadLedgers returns nothing), every such entry appears as a mismatch, producing false positives for valid stores after an upgrade. This makes the new verification unsafe for any deployment that has existing archived chapters.

**Suggestion:** After reading the baseline, check whether it contains a LedgerInfo section (or an equivalent format sentinel). If absent, treat the baseline as a legacy checkpoint and either skip the schema/account-type projection checks for archived chapters entirely, or re-seed the expected maps from the baseline data before running comparisons.

File: internal/application/check/checker.go [hunk 17]
```diff
@@ -1935,6 +1936,335 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	if err != nil {
+		return fmt.Errorf("reading ledgers for schema verification: %w", err)
+	}
+
+	ledgers, err := cursor.Collect(ledgerCursor)
+	if err != nil {
+		return fmt.Errorf("collecting ledgers for schema verification: %w", err)
+	}
+
+	for _, info := range ledgers {
```
Comment (NumaryBot) [inline internal/application/check/checker.go:1956]: 🟠 **[major] Missing LedgerInfo projections in stored ledger list not detected**

If a live ledger's LedgerInfo entry is absent from Pebble while the audit log still contains its CreateLedger event, the expected map contains that ledger but the loop over stored ledgers never visits it. No schema or account-type mismatch is emitted, creating a blind spot in the integrity verification the PR is meant to provide.

**Suggestion:** After iterating over stored ledgers, iterate over the expected maps and report any ledger present in the expected set but absent from the stored ledger list as a missing LedgerInfo projection mismatch.

File: internal/application/check/checker.go [hunk 18]
```diff
@@ -1935,6 +1936,335 @@ func verifySealingHash(p *commonpb.Chapter, callback func(*servicepb.CheckStoreE
...
+	ledgerCursor, err := query.ReadLedgers(ctx, reader)
+	if err != nil {
+		return fmt.Errorf("reading ledgers for presence verification: %w", err)
+	}
+
+	ledgers, err := cursor.Collect(ledgerCursor)
+	if err != nil {
+		return fmt.Errorf("collecting ledgers for presence verification: %w", err)
+	}
+
```
Comment (NumaryBot) [inline internal/application/check/checker.go:2120]: 🟠 **[major] Soft-deleted LedgerInfo entries not treated as missing during projection check**

If a live ledger's LedgerInfo projection is corrupted such that `DeletedAt` is set, `compareSchema` and `compareAccountTypes` both skip it as a tombstone, but the presence check still records the name as stored. An audit-live ledger with a soft-deleted LedgerInfo therefore produces no MISSING_LEDGER, schema, or account-type mismatch, creating a blind spot for this specific tampering shape.

**Suggestion:** Only count a LedgerInfo entry as satisfying the knownLedgers check when `DeletedAt` is not set. Treat soft-deleted entries as absent so they trigger the same mismatch path as a fully missing entry.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 3 new inline findings.

Summary: https://github.com/formancehq/ledger/pull/1555#issuecomment-4928117739

File: internal/application/check/checker.go [hunk 19]
```diff
@@ -230,58 +242,50 @@ func (c *Checker) Check(ctx context.Context, callback func(*servicepb.CheckStore
...
+		// from the live store. The audit hash chain — the real integrity guarantee
+		// — was already verified above. Expected after a checkpoint-only restore;
+		// warn and skip the entry-by-entry verification.
+		if baselineDB == nil {
+			c.logger.Error("no baseline checkpoint available for archived state comparison; skipping entry-by-entry verification")
 
-			if entry.Value.GetRevertedByTransaction() != 0 {
-				trackTxID(ledgerRevertedTxIDs, tk.LedgerName, tk.ID)
-			}
+			return nil
```
Comment (NumaryBot) [inline internal/application/check/checker.go:277]: 🟠 **[major] Early return skips post-archive log checks when no local baseline exists**

When an archived store has no local baseline checkpoint (expected after checkpoint-only restore), the early return exits before the log iterator runs. Previously, post-archive log projection checks — sequence gaps, malformed logs, reversion invariant violations, exclusion-record mismatches — still executed; now those corruptions pass silently with only the audit hash chain verified.

**Suggestion:** Restructure the control flow so that log iteration and its associated projection checks run regardless of whether a local baseline checkpoint is present. Only skip the baseline-dependent schema/account-type comparisons when the baseline is absent, rather than returning early from the entire check path.

File: internal/application/check/checker.go [hunk 20]
```diff
@@ -230,58 +242,50 @@ func (c *Checker) Check(ctx context.Context, callback func(*servicepb.CheckStore
...
 		}
 
-		if err := txIter.Close(); err != nil {
-			return fmt.Errorf("closing tx streaming iter: %w", err)
-		}
+		// The pre-archive live ledger set, schema, and account types were declared
+		// in now-purged logs; seed them from the boundary-time baseline so the
+		// post-archive replay applies its delta on top.
+		c.seedExpectedFromBaseline(ctx, baselineDB, knownLedgers, expectedSchemas, rawLedgerTypes, ledgerAccountTypes)
```
Comment (NumaryBot) [inline internal/application/check/checker.go:283]: 🟠 **[major] Legacy baselines without LedgerInfo cause false-positive mismatches on archived stores**
<sub>reported by NumaryBot, codex</sub>

Baseline checkpoints created before this PR contain only attribute zones and no LedgerInfo section. When such a baseline is opened, `baselineDB != nil` is still true, so `compareSchema` and `compareAccountTypes` run. Because schemas and account types declared before the archive point are absent from the expected maps (ReadLedgers returns nothing from legacy baselines), every such entry appears as a mismatch, producing false positives for valid stores after an upgrade. This makes the new verification unsafe for any deployment that has existing archived chapters.

**Suggestion:** After reading the baseline, detect whether it contains a LedgerInfo section (or equivalent sentinel). If absent, treat the baseline as a legacy checkpoint and either skip the schema/account-type projection checks for archived chapters entirely, or re-seed the expected maps from the baseline data before running comparisons.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 2 new inline findings.

Summary: https://github.com/formancehq/ledger/pull/1555#issuecomment-4928117739

File: internal/application/check/checker.go (hunk 19 above)
Comment (Azorlogh) [reply internal/application/check/checker.go:277]: Keeping the current behaviour deliberately. Running the log iteration and skipping only the baseline-dependent comparisons overstates what can safely run without a baseline under archiving:

- `UNKNOWN_LEDGER` reads `knownLedgers`, which under archiving is seeded from the baseline (pre-archive `CreateLedger` logs are purged). With no baseline it is empty, so the gate would flag every post-archive `Apply` — a false-positive flood.
- The reversion-invariant checks read `ledgerKnownTxIDs` / `ledgerRevertedTxIDs`, also baseline-seeded, so they would likewise false-fire.
- The exclusion-record compare builds its audit side from a replay that lacks every pre-archive order, so it would be partial and false-fire too.

Only `SEQUENCE_GAP` and malformed-log detection are cleanly baseline-independent, and a real sequence gap already breaks the audit hash chain (`HASH_MISMATCH`), which is still verified on this path. So the unique coverage lost in the (transient, post-restore) no-baseline case is minimal.

The overriding invariant is that the checker never seeds any state from the live store; without a baseline there is no independent source, so the correct behaviour is to skip entry-by-entry verification and surface it at error level rather than run partial checks that would false-fire. Resolving as intended.

**Comment by gfyrag** [comment]
Re-reviewed current head `c4943ba702ffce4a926fa5111e428916dd09cc39`.

Remaining findings:

1. High: legacy baselines are still trusted as current-format.

`Check` still trusts any opened baseline as current-format. Legacy baselines created before this PR contain no `LedgerInfo` rows, so `seedExpectedFromBaseline` leaves `knownLedgers`, schemas, and account types empty without telling `Check`. A valid archived store upgraded with an existing attributes-only baseline can then report `UNKNOWN_LEDGER`, `UNAUDITED_LEDGER`, `SCHEMA_MISMATCH`, or `ACCOUNT_TYPE_MISMATCH` for pre-archive ledgers. This was already raised by bot/codex; despite resolved threads, it is still open in the current code.

2. High: baseline `LedgerInfo` is used as expected archived state but is not seal-bound.

The baseline now copies `LedgerInfo`, and the checker uses it as expected archived state. But the chapter seal’s `state_hash` only covers volume, metadata, and transaction attributes, not `LedgerInfo`. For pre-archive schema/account-type declarations, the checker is comparing live `LedgerInfo` against another unverified projection, not against the audit chain or seal-bound state. That leaves a false-negative path if the baseline is stale/tampered consistently with live `LedgerInfo`.

3. Medium: archived stores with no local baseline still return success after only logging server-side.

`Check` returns nil on archived stores with no local baseline after logging. The `CheckStore` client receives no event indicating partial verification, and the log iterator below it never runs. Since `verifyAuditHashChain` hashes `AuditEntry`/`AuditItem` records, not the `SubColdLog` projection rows, this still skips some surviving log/projection checks in the no-baseline restore path.

4. Low: checker docs are stale.

`docs/technical/architecture/subsystems/checker/checker.md` still lists only the old passes/error enum through `INDEX_MISMATCH`, omitting schema, account types, missing ledger, and unaudited ledger.

What looks resolved in code:

- Initial account types from `CreateLedger`.
- Missing `LedgerInfo`.
- Soft-deleted `LedgerInfo`.
- Unaudited live `LedgerInfo`.
- The #1554 base contains the full `CreateLedger` reconstruction, `SubAttrLedger` dual-write, and account-type replay callbacks.

Note: PR #1554 still has restore-side gaps for `DeleteLedger`, `PromoteLedger`, and `UpdatedDefaultEnforcementMode`; this PR only partially intersects that via ledger presence checks and does not verify mode/mirror/default enforcement.

Checks run:

- `git diff --check 251cbc384519b32fac52b704fcac96d7020622e6..c4943ba702ffce4a926fa5111e428916dd09cc39`
- `git diff --check origin/fix/rebuilddelta-schema-2...c4943ba702ffce4a926fa5111e428916dd09cc39`
- `git merge-tree origin/fix/rebuilddelta-schema-2 c4943ba702ffce4a926fa5111e428916dd09cc39`
- `GOROOT= go test -count=1 ./internal/application/check ./internal/infra/backup`

**Comment by Azorlogh** [comment]
Folded into #1554 (commit 2d0f2bef4) following the review discussion there — the checker passes now land alongside the RebuildDelta fixes they verify, extended with `compareReferences` and `compareBoundaries`.
